data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,650 @@
+            },
+        }
+    }
+}
+
+impl Clone for ConstString {
+    fn clone(&self) -> Self {
+        if self.is_inlined() {
+            Self {
+                inlined: *self.inlined(),
+            }
+        } else {
+            Self {
+                boxed: self.boxed().clone(),
+            }
+        }
+    }
+}
+
+impl Drop for ConstString {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        if !self.is_inlined() {
+            // SAFETY: safe because`is_inlined` == `false`.
+            unsafe {
+                ManuallyDrop::drop(&mut self.boxed);
+            }
+        }
+    }
+}
+
+impl Serialize for ConstString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self)
+    }
+}
+
+impl<'de> Deserialize<'de> for ConstString {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        deserializer.deserialize_str(ConstStringVisitor)
+    }
+}
+
+struct ConstStringVisitor;
+
+impl<'de> Visitor<'de> for ConstStringVisitor {
+    type Value = ConstString;
+
+    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("a string")
+    }
+
+    fn visit_str<E: Error>(self, v: &str) -> Result<Self::Value, E> {
+        Ok(v.into())
+    }
+
+    fn visit_string<E: Error>(self, v: String) -> Result<Self::Value, E> {
+        Ok(v.into())
+    }
+}
+
+impl WrapperTypeEncode for ConstString {}
+
+impl WrapperTypeDecode for ConstString {
+    type Wrapped = String;
+}
+
+impl IntoSchema for ConstString {
+    fn type_name() -> String {
+        String::type_name()
+    }
+    fn schema(map: &mut MetaMap) {
+        String::schema(map);
+    }
+}
+
+#[derive(DebugCustom)]
+#[debug(fmt = "{:?}", "&**self")]
+#[repr(C)]
+struct BoxedString {
+    #[cfg(target_endian = "little")]
+    ptr: *mut u8,
+    len: usize,
+    #[cfg(target_endian = "big")]
+    ptr: *mut u8,
+}
+
+impl BoxedString {
+    #[inline]
+    const fn len(&self) -> usize {
+        self.len
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        // SAFETY: created from `Box<[u8]>`.
+        unsafe { from_raw_parts(self.ptr, self.len) }
+    }
+}
+
+impl AsRef<str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8
+        unsafe { from_utf8_unchecked(self.as_bytes()) }
+    }
+}
+
+impl Deref for BoxedString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Clone for BoxedString {
+    /// Properly clone `BoxedString` into new allocation.
+    fn clone(&self) -> Self {
+        let ptr = Box::into_raw(self.as_bytes().to_owned().into_boxed_slice()).cast::<u8>();
+        let len = self.len;
+        Self { ptr, len }
+    }
+}
+
+impl From<&str> for BoxedString {
```

Comment:
implement `FromStr` instead :)

Reply:
Technically, this is an infallible conversion, so we should have both.

Reply:
@mversic, i think `FromStr` exist in parsing contexts and also have notion of failure.
In our case it more of the conversion and with `From` + `TryForm` it's possible to create generic implementation `From` for `ConstString`.
Also creation of `BoxedString` from string can't fail.
However I could implement it in the same way as `String` [did](https://doc.rust-lang.org/src/alloc/string.rs.html#2351).

Reply:
ok, I missed that it's infallible, in which case there shouldn't be an implementation of `From<&str>`. Rather one should turn `str` into `String` and call `From<String>`

Reply:
Std has [`impl<'_> From<&'_ str> for String`](https://doc.rust-lang.org/std/convert/trait.From.html#impl-From%3C%26%27_%20str%3E-3) , so i think it would be suitable to implement it for `BoxedString` (and `ConstString`) too. 

I we want to call `parse` we can implement `FromStr` on the top level for the `ConstString`.